lib/redis-session-store.rb

```diff
@@ -134,73 +123,68 @@ def decode(data)
     USE_INDIFFERENT_ACCESS ? session.with_indifferent_access : session
   end
 
-  def set_session(env, sid, session_data, options = nil)
-    expiry = get_expiry(env, options)
-    with_redis do |redis|
-      if expiry
-        redis.setex(prefixed(sid), expiry, encode(session_data))
-      else
-        redis.set(prefixed(sid), encode(session_data))
-      end
+  def write_session(req, sid, session_data, options = nil)
+    return false unless sid
+
+    if write_fallback
+      key = prefixed_fallback(sid)
+    else
+      key = prefixed(sid)
     end
-    sid
-  rescue Errno::ECONNREFUSED, Redis::CannotConnectError => e
-    on_redis_down.call(e, env, sid) if on_redis_down
-    false
-  end
-  alias write_session set_session
+    return false unless key
 
-  def get_expiry(env, options)
-    session_storage_options = options || env.fetch(ENV_SESSION_OPTIONS_KEY, {})
-    session_storage_options[:ttl] || session_storage_options[:expire_after]
+    with_redis_connection(false) do |redis_connection|
+      redis_connection.set(key, encode(session_data), ex: ttl(default_redis_ttl, options[:expire_after]))
+      sid
+    end
   end
 
   def encode(session_data)
     serializer.dump(session_data)
   end
 
-  def destroy_session(env, sid, options)
-    destroy_session_from_sid(sid, (options || {}).to_hash.merge(env: env))
+  def delete_session(req, sid, options)
+    with_redis_connection do |redis_connection|
+      delete_session_from_redis(redis_connection, sid, options)
+    end
   end
-  alias delete_session destroy_session
 
-  def destroy(env)
-    if env['rack.request.cookie_hash'] &&
-       (sid = env['rack.request.cookie_hash'][key])
-      destroy_session_from_sid(sid, drop: true, env: env)
+  def delete_session_from_redis(redis_connection, sid, options)
+    if write_fallback
+      key = prefixed_fallback(sid)
+    else
+      key = prefixed(sid)
     end
-    false
+    redis_connection.del(key)
+    create_sid_with_empty_session(redis_connection) unless options[:drop]
   end
 
-  def destroy_session_from_sid(sid, options = {})
-    with_redis do |redis|
-      redis.del(prefixed(sid))
-    end
-    (options || {})[:drop] ? nil : generate_sid
-  rescue Errno::ECONNREFUSED, Redis::CannotConnectError => e
-    on_redis_down.call(e, options[:env] || {}, sid) if on_redis_down
+  def ttl(ttl, expire_after)
+    expire_after || ttl
   end
 
   def determine_serializer(serializer)
     serializer ||= :marshal
     case serializer
     when :marshal then Marshal
     when :json    then JsonSerializer
-    when :hybrid  then HybridSerializer
     else serializer
     end
   end
 
   # Consistent interface for a redis instance from a pool
   # @yield [Redis]
-  def with_redis
+  def with_redis_connection(default_value = nil)
```

Comment:
WDYT of calling it `default_rescue_value` and making it a keyword argument? it makes it sounds like it would be the default return instead of the yielded block

```suggestion
  def with_redis_connection(default_rescue_value: nil)
```

Reply:
agree, renamed it in 6444fe7